### PR TITLE
[WIP] [Speculative Decoding] Support Draft-Model Speculative Decoding on Metal

### PR DESCRIPTION
**Pull request by davidgao7**
# Summary

This PR introduces the foundational architecture for Speculative Decoding on the Metal paged path. It implements a Widened Scorer approach that leverages MLX's unified varlen kernels to verify multiple draft tokens in a single, efficient forward pass.

# Key Changes

* **MetalSpecDecodeWorker**: New orchestrator that manages the Propose-Score-Verify-Rewind loop and stashes speculative data for the asynchronous Paged Attention path.
* Full Worker Delegation: Implemented missing *WorkerBase* RPC methods (*take_draft_token_ids*, *get_supported_tasks*, health checks) to ensure full compatibility with the vLLM EngineCore and benchmarking tools.
* MLX-Native **MetalDraftProposer**: A functional proposer using a secondary model lifecycle that leverages MLX lazy evaluation to build draft sequences without GPU stalls.
* Vectorized **MetalRejectionSampler**: A parallelized rejection sampler with the "residual fix" (*mx.maximum*) to ensure mathematical parity with sequential generation.
* Atomic KV Rewind: Refactored *ModelRunner* to support *truncate_cache* with bonus token commitment, ensuring the KV cache and sequence length tracking stay in sync after rejections.
* Vectorized Logit Extraction: Updated *sampling_batch.py* to extract and return "Widened Pipe" verification logits for the entire speculative window.
* macOS Stability: Automated loopback networking defaults (*lo0* / *VLLM_HOST_IP*) to prevent Gloo distributed hangs on Apple Silicon.

# Performance Results (M4 Max)

* **Sequential (1 token)**: 4.92ms

* **Widened Verification (5 tokens)**: 6.65ms

* **3.7x** hardware-level throughput increase during verification.

# Edge Case Handling
To prevent [state-sync issues](https://github.com/vllm-project/vllm/issues/34650), this PR implements:
1. Strict Boundary Control: The *MetalModelRunner* now decrementally updates sequence tracking, forcing the GPU to ignore rejected tokens in subsequent passes.
2. Verification Stashing: Added *_pending_spec_data* in the worker to handle the handshake between the asynchronous Scorer pass and the verification phase.

# Validation & Test Plan
Hardware:Apple MacBook Pro 16 (2024) with M4 Max (36GB RAM):
Verified using the vLLM CLI (v0.19.0) and internal integration tests. Hardware-saturated benchmarks confirm the Scorer's efficiency gain.

1. Baseline Measurement (Target Model Only):

```bash
vllm bench latency --model mlx-community/Qwen2.5-0.5B-4bit --num-iters 10
```

2. Speculative Measurement (Target + Draft):

```bash
vllm bench latency \
  --model mlx-community/Qwen2.5-0.5B-4bit \
  --speculative-model mlx-community/Qwen2.5-0.5B-4bit \
  --num-speculative-tokens 5 \
  --num-iters 10
```

Metrics of Interest:

- `ITL` (Inter-Token Latency): Expected to decrease by 30-50% on the M4 Max.
- `TPOT` (Time Per Output Token): The primary success metric for the MQA Scorer's efficiency.

### CODE REVIEW

**Comment by davidgao7** [comment]
Here is the result I tested locally:

Scorer Verification Benchmark:
* Sequential Generation (avg 1 token): 4.92ms
* Widened Verification (avg 5 tokens): 6.65ms
* Hardware Efficiency Gain: 3.7x more efficient per token verified than sequential generation.

```bash
VLLM_USE_MODELSCOPE=True pytest tests/test_spec_decode_proposer.py -s -m slow
```

```log
> VLLM_USE_MODELSCOPE=True pytest tests/test_spec_decode_proposer.py -s -m slow
================================================================================================================ test session starts =================================================================================================================
platform darwin -- Python 3.13.13, pytest-9.0.2, pluggy-1.6.0
rootdir: /Users/tengjungao/Downloads/vllm-metal
configfile: pyproject.toml
plugins: anyio-4.12.1, asyncio-1.3.0
asyncio: mode=Mode.AUTO, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collecting ... INFO 05-04 18:21:23 [__init__.py:44] Available plugins for group vllm.platform_plugins:
INFO 05-04 18:21:23 [__init__.py:46] - metal -> vllm_metal:register
INFO 05-04 18:21:23 [__init__.py:49] All plugins in this group will be loaded. Set `VLLM_PLUGINS` to control which plugins to load.
INFO 05-04 18:21:24 [__init__.py:239] Platform plugin metal is activated
collected 2 items

tests/test_spec_decode_proposer.py INFO 05-04 18:21:26 [utils.py:41] Downloading model mlx-community/Qwen2.5-0.5B-4bit from ModelScope...
Downloading Model from https://www.modelscope.cn to directory: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2.5-0.5B-4bit
2026-05-04 18:21:34,235 - modelscope - INFO - Target directory already exists, skipping creation.
INFO 05-04 18:21:34 [utils.py:43] Model downloaded to /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:34 [model.py:549] Resolved architecture: Qwen2ForCausalLM
WARNING 05-04 18:21:34 [model.py:2016] Casting torch.bfloat16 to torch.float16.
INFO 05-04 18:21:34 [model.py:1678] Using max model len 32768
INFO 05-04 18:21:34 [scheduler.py:238] Chunked prefill is enabled with max_num_batched_tokens=2048.
INFO 05-04 18:21:34 [model.py:549] Resolved architecture: Qwen2ForCausalLM
WARNING 05-04 18:21:34 [model.py:2016] Casting torch.bfloat16 to torch.float16.
INFO 05-04 18:21:34 [model.py:1678] Using max model len 32768
WARNING 05-04 18:21:34 [vllm.py:762] Async scheduling not supported with draft_model-based speculative decoding and will be disabled.
INFO 05-04 18:21:34 [vllm.py:790] Asynchronous scheduling is disabled.
INFO 05-04 18:21:34 [importing.py:68] Triton not installed or not compatible; certain GPU-related functions will not be available.
INFO 05-04 18:21:34 [platform.py:264] Metal: chunked prefill enabled (paged attention), max_num_batched_tokens=2048
INFO 05-04 18:21:34 [platform.py:329] Metal memory: 38.7GB total, 18.4GB available
INFO 05-04 18:21:34 [proposer.py:83] Loading draft model from: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:34 [proposer.py:57] Calling ModelLifecycle.load() for /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit...
INFO 05-04 18:21:34 [model_lifecycle.py:175] Loading model: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit (VLM: False)
INFO 05-04 18:21:34 [model_lifecycle.py:207] Model loaded in 0.35s: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:34 [proposer.py:59] ModelLifecycle.load() completed.
INFO 05-04 18:21:34 [proposer.py:57] Calling ModelLifecycle.load() for /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit...
INFO 05-04 18:21:34 [model_lifecycle.py:175] Loading model: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit (VLM: False)
INFO 05-04 18:21:34 [model_lifecycle.py:182] Model loaded from cache in 0.000s: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:34 [proposer.py:59] ModelLifecycle.load() completed.
INFO 05-04 18:21:34 [proposer.py:109] Initializing draft cache for model of type <class 'mlx_lm.models.qwen2.Model'>
INFO 05-04 18:21:34 [proposer.py:114] Running prefill for 5 tokens...
INFO 05-04 18:21:34 [proposer.py:127] Building lazy graph for 5 steps...
INFO 05-04 18:21:34 [proposer.py:146] Graph building complete. Returning to caller.

========================================
--- Scorer Verification Benchmark ---
Scorer Sequential (avg 1 token): 4.92ms
Scorer Verification (avg 5 tokens): 6.65ms

VERDICT: Scorer is 3.7x more efficient via Speculation!
========================================

.INFO 05-04 18:21:35 [utils.py:41] Downloading model mlx-community/Qwen2.5-0.5B-4bit from ModelScope...
Downloading Model from https://www.modelscope.cn to directory: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2.5-0.5B-4bit
2026-05-04 18:21:41,518 - modelscope - INFO - Target directory already exists, skipping creation.
INFO 05-04 18:21:41 [utils.py:43] Model downloaded to /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:41 [model.py:549] Resolved architecture: Qwen2ForCausalLM
WARNING 05-04 18:21:41 [model.py:2016] Casting torch.bfloat16 to torch.float16.
INFO 05-04 18:21:41 [model.py:1678] Using max model len 32768
INFO 05-04 18:21:41 [model.py:549] Resolved architecture: Qwen2ForCausalLM
WARNING 05-04 18:21:41 [model.py:2016] Casting torch.bfloat16 to torch.float16.
INFO 05-04 18:21:41 [model.py:1678] Using max model len 32768
INFO 05-04 18:21:41 [platform.py:264] Metal: chunked prefill enabled (paged attention), max_num_batched_tokens=2048
INFO 05-04 18:21:41 [platform.py:329] Metal memory: 38.7GB total, 17.9GB available
INFO 05-04 18:21:41 [worker.py:59] MetalSpecDecodeWorker initialized.
INFO 05-04 18:21:41 [worker.py:128] MLX device set to: Device(gpu, 0)
mx.metal.device_info is deprecated and will be removed in a future version. Use mx.device_info instead.
INFO 05-04 18:21:41 [utils.py:73] Set Metal wired_limit to 27.0 GB
INFO 05-04 18:21:41 [worker.py:136] PyTorch device set to: mps
INFO 05-04 18:21:41 [parallel_state.py:1400] world_size=1 rank=0 local_rank=0 distributed_init_method=tcp://localhost:12345 backend=gloo
INFO 05-04 18:21:41 [parallel_state.py:1716] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, PCP rank 0, TP rank 0, EP rank N/A, EPLB rank N/A
WARNING 05-04 18:21:41 [__init__.py:206] min_p and logit_bias parameters won't work with speculative decoding.
INFO 05-04 18:21:41 [model_lifecycle.py:175] Loading model: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit (VLM: False)
INFO 05-04 18:21:41 [model_lifecycle.py:182] Model loaded from cache in 0.000s: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:41 [proposer.py:83] Loading draft model from: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:41 [proposer.py:57] Calling ModelLifecycle.load() for /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit...
INFO 05-04 18:21:41 [model_lifecycle.py:175] Loading model: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit (VLM: False)
INFO 05-04 18:21:41 [model_lifecycle.py:182] Model loaded from cache in 0.000s: /Users/tengjungao/.cache/modelscope/hub/models/mlx-community/Qwen2___5-0___5B-4bit
INFO 05-04 18:21:41 [proposer.py:59] ModelLifecycle.load() completed.
INFO 05-04 18:21:41 [cache_policy.py:720] Paged attention: VLLM_METAL_MEMORY_FRACTION=auto, defaulting to 0.90 for paged path
INFO 05-04 18:21:41 [cache_policy.py:568] Paged attention memory breakdown: metal_limit=28.99GB, fraction=0.90, usable_metal=26.09GB, model_memory=0.28GB, overhead=0.78GB, kv_budget=25.03GB, per_block_bytes=196608, num_blocks=127317, max_tokens_cached=2037072
INFO 05-04 18:21:44 [cache.py:166] KV cache: 25031.5 MB (24 layers, 127317 blocks, 16 tokens/block)
INFO 05-04 18:21:45 [cache_policy.py:601] Paged attention enabled: 24 layers patched, 127317 blocks allocated (block_size=16, mla=False, turboquant=False, k_quant=N/A)
INFO 05-04 18:21:45 [cache_policy.py:640] Paged attention: reporting MPS cache capacity (127317 blocks × 196608 bytes = 25.03 GB)
INFO 05-04 18:21:45 [proposer.py:109] Initializing draft cache for model of type <class 'mlx_lm.models.qwen2.Model'>
INFO 05-04 18:21:45 [proposer.py:114] Running prefill for 4 tokens...
INFO 05-04 18:21:45 [proposer.py:127] Building lazy graph for 5 steps...
INFO 05-04 18:21:45 [proposer.py:146] Graph building complete. Returning to caller.
INFO 05-04 18:21:45 [__init__.py:230] Native paged-attention Metal kernels loaded
.

================================================================================================================== warnings summary ==================================================================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

tests/test_spec_decode_proposer.py: 14 warnings
  /Users/tengjungao/Downloads/vllm-metal/.venv/lib/python3.13/site-packages/torch/jit/_script.py:362: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================================================== 2 passed, 16 warnings in 22.67s ===========================================================================================================
<sys>:0: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

```

**Comment by davidgao7** [comment]
Hey team, I've got the core speculative decoding (the 'Wide Pipe' architecture) finished for Metal.

Locally, I'm seeing a 3.7x speedup on my M4 Max (4.92ms vs 6.65ms for a 5-token verification), which is huge for local inference.

Since this touches about 10 files to get the full loop working (worker orchestration, MLX sampling math, and KV rewinding), I’ve grouped the changes in the description to make it easier to skim. I’m currently keeping the [WIP] tag because I’m starting on Stateful Cache to shave off another ~3ms of overhead, but the core logic here is ready for a look if anyone wants to start early!

**Review by WindChimeRan — COMMENTED** [review]
Could you please check #308 and #349 to avoid duplicated work? 

Please talk to @CAICAIIs . Maybe you guys can collaborate and distribute the work.


* `macOS Stability: Automated loopback networking defaults (lo0 / VLLM_HOST_IP) to prevent Gloo distributed hangs on Apple Silicon.` this is irrelevant to Spec Decoding. Please break these to dedicated, smaller PR
* `3.7x speedup` is misleading. The exp setting is using the same model for both target and draft model. I think negative speedup is expected?

**Comment by WindChimeRan** [comment]
@davidgao7 Hi, do you have any updates on this PR?

**Comment by WindChimeRan** [comment]
will open a new PR for it soon.
